primitives/maintenance-mode-primitives/src/hooks.rs

```diff
@@ -0,0 +1,68 @@
+// Copyright Moonsong Labs
+// This file is part of Moonkit.
+
+// Moonkit is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonkit is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonkit.  If not, see <http://www.gnu.org/licenses/>.
+
+use core::marker::PhantomData;
+use frame_support::{
+	traits::{OffchainWorker, OnFinalize, OnIdle, OnInitialize, OnRuntimeUpgrade},
+	weights::Weight,
+};
+
+#[cfg(feature = "try-runtime")]
+use sp_runtime::DispatchError;
+
+pub type BlockNumber = u32;
+
+/// The hooks we want to run in Maintenance Mode
+pub struct MaintenanceHooks<T>(PhantomData<T>);
+impl<T: OnInitialize<BlockNumber>> OnInitialize<BlockNumber> for MaintenanceHooks<T> {
+	fn on_initialize(n: BlockNumber) -> Weight {
+		T::on_initialize(n)
+	}
+}
+
+// We override onIdle for xcmQueue and dmpQueue pallets to not process messages inside it
```

Comment:
It doesn't work because the on_idle implementation shuffle the pallets order, so we should probably remove the whole MaintenanceHooks as it's useless.
For DMP, with polkadot 1.6 the dmp queue will be replaced by the message queue that can be paused properly by setting a storage item, so it will not be an issue